File: jobhunter/letters.py
"""
Cover letters.

Με ANTHROPIC_API_KEY → γράφει ο Claude, προσαρμοσμένο στο συγκεκριμένο job.
Χωρίς key → solid template fallback, ώστε το εργαλείο να δουλεύει ούτως ή άλλως.

Το Anthropic API καλείται με σκέτο urllib — κανένα SDK, κανένα pip install.
"""
from __future__ import annotations

import json
import re
import urllib.error
import urllib.request

from .config import Settings
from .profile import Profile

API_URL = "https://api.anthropic.com/v1/messages"
API_VERSION = "2023-06-01"

SYSTEM_PROMPT = """You write high-converting cover letters for job seekers.

Rules:
1. Open with one specific sentence about THIS company/role — never a generic hook.
2. Three concrete value propositions, each anchored to a verifiable fact from the CV.
3. Mirror the vocabulary of the job description.
4. Use metrics wherever the CV supplies them. Never invent experience.
5. 220-320 words. Recruiters skim for 7 seconds.
6. No clichés: "hard worker", "team player", "passionate", "I believe".
7. Plain text only — no markdown, no headers, no placeholders like [Company].
8. Write in the requested language, natively (not translated-sounding).
9. Start with an appropriate greeting, end with the candidate's name."""

LANG_NAMES = {"en": "English", "el": "Greek", "es": "Spanish",
              "de": "German", "fr": "French", "it": "Italian"}


class LetterError(RuntimeError):
    pass


def generate(job: dict, profile: Profile, settings: Settings,
             language: str | None = None) -> tuple[str, str]:
    """Returns (content, model_used). Πέφτει σε template αν δεν υπάρχει key."""
    lang = language or profile.default_language
    cv_text = profile.cv_text(lang)

    if not settings.has_ai or not cv_text:
        return _template(job, profile, lang), "template"

    try:
        return _claude(job, profile, settings, lang, cv_text), settings.claude_model
    except (urllib.error.URLError, LetterError, KeyError, ValueError, TimeoutError) as e:
        return _template(job, profile, lang) + f"\n\n[fallback: AI generation failed — {e}]", "template"


def _claude(job: dict, profile: Profile, settings: Settings, lang: str, cv_text: str) -> str:
    user_prompt = (
        f"Write a cover letter in {LANG_NAMES.get(lang, 'English')}.\n\n"
        f"COMPANY: {job.get('company')}\n"
        f"ROLE: {job.get('title')}\n"
        f"LOCATION: {job.get('location') or 'n/a'}\n\n"
        f"JOB DESCRIPTION:\n{(job.get('description') or '')[:4000]}\n\n"
        f"CANDIDATE CV:\n{cv_text[:8000]}\n\n"
        "Return only the letter."
    )

    payload = {
        "model": settings.claude_model,
        "max_tokens": 1200,
        "system": [{
            "type": "text",
            "text": SYSTEM_PROMPT,
            "cache_control": {"type": "ephemeral"},
        }],
        "messages": [{"role": "user", "content": user_prompt}],
    }

    req = urllib.request.Request(
        API_URL,
        data=json.dumps(payload).encode("utf-8"),
        headers={
            "content-type": "application/json",
            "x-api-key": settings.anthropic_api_key,
            "anthropic-version": API_VERSION,
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=90) as resp:
            data = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        detail = e.read().decode("utf-8", errors="replace")[:400]
        raise LetterError(f"Anthropic HTTP {e.code}: {detail}") from e

    blocks = [b.get("text", "") for b in data.get("content", []) if b.get("type") == "text"]
    text = "\n".join(blocks).strip()
    if not text:
        raise LetterError("κενή απάντηση από το API")
    return text


# ════════════════════════════════════════════════════════════════
# Template fallback
# ════════════════════════════════════════════════════════════════
# Τα bullets σε markdown CV σπάνε σε πολλές γραμμές — τα ενώνουμε ξανά.
_BULLET_RE = re.compile(r"^[ \t]*[-*•][ \t]+(.+?)(?=\n[ \t]*[-*•][ \t]|\n[ \t]*\n|\n#|\Z)",
                        re.M | re.S)


def _cv_highlights(profile: Profile, lang: str, n: int = 3) -> list[str]:
    text = profile.cv_text(lang, limit=6000)
    bullets = []
    for raw in _BULLET_RE.findall(text):
        clean = re.sub(r"\s+", " ", re.sub(r"\*\*|\*|`|_", "", raw)).strip(" .")
        if 40 <= len(clean) <= 220:
            bullets.append(clean)
    return bullets[:n]


def _template(job: dict, profile: Profile, lang: str) -> str:
    company = job.get("company", "your company")
    title = job.get("title", "the role")
    highlights = _cv_highlights(profile, lang)
    name = profile.name or "—"

    if lang == "el":
        body = [
            f"Αξιότιμοι κύριοι/κυρίες της {company},",
            "",
            f"Γράφω για τη θέση {title}. Το προφίλ μου συνδυάζει άμεση εμπειρία "
            f"σε πελατοκεντρικούς ρόλους με πρακτική γνώση του κλάδου σας.",
            "",
        ]
        body += [f"• {h}" for h in highlights] or ["• Δείτε το βιογραφικό μου για αναλυτικά στοιχεία."]
        body += [
            "",
            f"Θα χαρώ πολύ να συζητήσουμε πώς μπορώ να συνεισφέρω στην ομάδα της {company}. "
            "Είμαι διαθέσιμος για συνέντευξη άμεσα.",
            "",
            "Με εκτίμηση,",
            name,
        ]
    else:
        body = [
            f"Dear {company} hiring team,",
            "",
            f"I am applying for the {title} position. My background combines hands-on "
            f"customer-facing experience with practical domain knowledge of your industry.",
            "",
        ]
        body += [f"• {h}" for h in highlights] or ["• Please see my CV for detailed background."]
        body += [
            "",
            f"I would welcome the chance to discuss how I can contribute to {company}. "
            "I am available for an interview immediately.",
            "",
            "Kind regards,",
            name,
        ]

    contact = " · ".join(x for x in (profile.email, profile.phone, profile.linkedin) if x)
    if contact:
        body += ["", contact]
    return "\n".join(body)

File: extension/lib/letters.js
/*
 * Cover letters. Χωρίς κλειδί → σοβαρό template. Με κλειδί Anthropic → Claude.
 * Το κλειδί είναι του χρήστη, μένει στο chrome.storage του και πάει μόνο στο
 * api.anthropic.com.
 */

const SYSTEM = `You write high-converting cover letters for job seekers.

Rules:
1. Open with one specific sentence about THIS company and role.
2. Three concrete value propositions, each anchored to a fact from the CV.
3. Mirror the vocabulary of the job description.
4. Never invent experience the CV does not support.
5. 220-320 words. Plain text only, no markdown, no placeholders.
6. Write natively in the requested language.
7. No clichés: "hard worker", "team player", "passionate".`;

const LANG_NAMES = { en: "English", el: "Greek", es: "Spanish", de: "German",
  fr: "French", it: "Italian", pt: "Portuguese", nl: "Dutch", pl: "Polish" };

export async function generateLetter(job, profile, cv) {
  const lang = profile.language || "en";
  const cvText = (cv && cv.text) || "";

  if (profile.anthropicKey && (cvText || profile.about)) {
    try {
      return { text: await viaClaude(job, profile, cvText, lang), model: "claude" };
    } catch (e) {
      return { text: template(job, profile, cvText, lang), model: "template", error: String(e.message || e) };
    }
  }
  return { text: template(job, profile, cvText, lang), model: "template" };
}

async function viaClaude(job, profile, cvText, lang) {
  const res = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-api-key": profile.anthropicKey,
      "anthropic-version": "2023-06-01",
      "anthropic-dangerous-direct-browser-access": "true",
    },
    body: JSON.stringify({
      model: "claude-haiku-4-5-20251001",
      max_tokens: 1200,
      system: [{ type: "text", text: SYSTEM, cache_control: { type: "ephemeral" } }],
      messages: [{
        role: "user",
        content: `Write a cover letter in ${LANG_NAMES[lang] || "English"}.

COMPANY: ${job.company}
ROLE: ${job.title}
LOCATION: ${job.location || "n/a"}

JOB DESCRIPTION:
${(job.description || "").slice(0, 4000)}

CANDIDATE CV:
${cvText.slice(0, 8000)}
${profile.about ? `
IN THE CANDIDATE'S OWN WORDS:
${profile.about.slice(0, 1500)}
` : ""}
Return only the letter.`,
      }],
    }),
  });

  if (!res.ok) throw new Error(`Anthropic ${res.status}: ${(await res.text()).slice(0, 200)}`);
  const data = await res.json();
  const text = (data.content || []).filter((b) => b.type === "text").map((b) => b.text).join("\n").trim();
  if (!text) throw new Error("empty response");
  return text;
}

/* ── Template fallback ────────────────────────────────────── */

// Επικεφαλίδες και στοιχεία επικοινωνίας δεν είναι επιτεύγματα.
const NOISE = /@|https?:|linkedin|github|curriculum vitae|^\s*resume\b|\+\d{2}|\[.*\]/i;

/**
 * Τα bullets ενός CV σπάνε σε πολλές γραμμές, οπότε τα ξαναενώνουμε πρώτα.
 * Προτιμώνται πάντα οι γραμμές με πραγματικό bullet — εκεί είναι η ουσία.
 */
function highlights(cvText, n = 3) {
  const bullets = [];
  const fallback = [];
  let current = null;
  let currentIsBullet = false;

  const flush = () => {
    if (!current) return;
    const line = current.replace(/\*\*|`|_/g, "").replace(/\s+/g, " ").trim().replace(/[.;,]$/, "");
    current = null;
    if (line.length < 45 || line.length > 220 || line.endsWith(":") || NOISE.test(line)) return;
    (currentIsBullet ? bullets : fallback).push(line);
  };

  for (const raw of cvText.split(/\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || line.startsWith("---")) { flush(); continue; }

    const bulletMatch = /^[-*•●▪]\s+(.*)$/.exec(line);
    if (bulletMatch) {
      flush();
      current = bulletMatch[1];
      currentIsBullet = true;
    } else if (current) {
      current += " " + line;                       // συνέχεια της ίδιας γραμμής
    } else {
      current = line;
      currentIsBullet = false;
    }
    if (bullets.length >= n) break;
  }
  flush();

  return (bullets.length ? bullets : fallback).slice(0, n);
}

const T = {
  en: (c, t, h, name, contact) => [
    `Dear ${c} hiring team,`, "",
    `I am applying for the ${t} position. My background lines up closely with what the role asks for.`, "",
    ...(h.length ? h.map((x) => `• ${x}`) : ["• Please see my attached CV for the full background."]),
    "", `I would welcome the chance to discuss how I can contribute to ${c}. I am available for an interview at short notice.`,
    "", "Kind regards,", name, contact,
  ],
  el: (c, t, h, name, contact) => [
    `Αξιότιμοι κύριοι/κυρίες της ${c},`, "",
    `Γράφω για τη θέση ${t}. Το προφίλ μου ταιριάζει άμεσα με όσα ζητά ο ρόλος.`, "",
    ...(h.length ? h.map((x) => `• ${x}`) : ["• Παρακαλώ δείτε το συνημμένο βιογραφικό μου."]),
    "", `Θα χαρώ να συζητήσουμε πώς μπορώ να συνεισφέρω στην ${c}. Είμαι διαθέσιμος/η για συνέντευξη άμεσα.`,
    "", "Με εκτίμηση,", name, contact,
  ],
};

function template(job, profile, cvText, lang) {
  const make = T[lang] || T.en;
  const contact = [profile.email, profile.phone, profile.linkedin].filter(Boolean).join(" · ");
  return make(job.company, job.title, highlights(cvText), profile.name || "", contact)
    .join("\n").replace(/\n{3,}/g, "\n\n").trim();
}
